File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        i = 0
        first = ""
        second = ""
        diff1 = 0
        diff2 = 0
        result = 0
        if x > y:
            first = "ab"
            second = "ba"
            diff1 = x
            diff2 = y
        else:
            first = "ba"
            second = "ab"
            diff1 = y
            diff2 = x
        
        stack = []

        while(i < len(s)):
            stack.append(s[i])
            if len(stack) > 1 and str(stack[-2] + stack[-1]) == first:
                result += diff1
                del stack[-1]
                del stack[-1]
            i = i + 1
        #print(''.join(stack), result)
        new_stack = []
        while stack :
            new_stack.append(stack.pop(0))
            if len(new_stack) > 1 and new_stack[-2] + new_stack[-1] == second:
                result += diff2
                new_stack.pop()
                new_stack.pop()
                
        return result




        
        

File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.rs
impl Solution {
    pub fn maximum_gain(s: String, x: i32, y: i32) -> i32 {
        let mut points = 0;
        let mut chars: Vec<char> = s.chars().collect();

        if x > y {
            points += Self::remove_all(&mut chars, 'a', 'b', x);
            points += Self::remove_all(&mut chars, 'b', 'a', y);
        } else {
            points += Self::remove_all(&mut chars, 'b', 'a', y);
            points += Self::remove_all(&mut chars, 'a', 'b', x);
        }

        points
    }

    fn remove_all(chars: &mut Vec<char>, first: char, second: char, score: i32) -> i32 {
        let mut stack = Vec::new();
        let mut points = 0;

        for &ch in chars.iter() {
            if let Some(&last) = stack.last() {
                if last == first && ch == second {
                    stack.pop();
                    points += score;
                    continue;
                }
            }
            stack.push(ch);
        }

        // Replace original vec with remaining characters for next removal round
        chars.clear();
        chars.extend(stack);
        points
    }
}
